### DISCUSSION TITLE
Open epub to location with epubcfi

### DISCUSSION THREAD

**Discussion by ondrejdyck** [johnfactotum/foliate#1011] [Q&A] [answered]
Is there a way to tell Foliate to open a book to a specific location using an epubcfi?

I'm trying to organize my highlights/notes using Obsidian and have made (started on) a script that goes through the .json files and dumps the highlights into the Obsidian vault as .md files.

What I would like to do is somehow create a clickable link in the .md file that opens the book with Foliate and scrolls to the proper location. I think I have read somewhere that we can get Obsidian to do dangerous things such as execute a script but the key piece of this will be passing the epubcfi information to Foliate.

A hacky way would be to overwrite the current location store in the .json file with the quote location but then one would loose their position in the book. Oh, well, hmm. I guess the position in the book would be changed anyway because you scrolled to a new location. Maybe this is the way?

**Comment by johnfactotum** [comment, marked as answer]
See #997.

> A hacky way would be to overwrite the current location store in the .json file with the quote location but then one would loose their position in the book. Oh, well, hmm. I guess the position in the book would be changed anyway because you scrolled to a new location.

Yeah, you would lose the location anyway, but you can store the location as a bookmark in the JSON file.

> Maybe this is the way?

It does have some advantages:

- Works right now with any version of Foliate.
- Gives you control over whether or how to save the last location.

Cons:

- Feels kind of hacky.
- Requires additional disk read/write.

**Reply by ondrejdyck** [reply]
Well, I can confirm that this does work. But it is even more hacky on the Obsidian side since what I ended up doing was create a code cell at the end of the note that you can run that runs a script that edits the .json file and then starts Foliate. This requires that you have the Code Snippets plugin installed.

The real problem with this approach is that you have to know where the epub file is located in order to tell Foliate to open it. So far I've not stumbled across a log of where Foliate stores the file locations but, since it has the library view, it must be writing this information down somewhere. Is there another .json file somewhere we can read this information?

EDIT: I spoke too soon. I see there is a folder called Library and `uri-store.json` that contains this information. This is fantastic!

**Reply by starise** [reply]
Do you mind to share your script/solution? I'm searching exactly for a way to import my epub annotations to Obsidian with links to open the proper location on Foliate. Thank you.

**Reply by ondrejdyck** [reply]
Starise, I put it on github here with a brief description of how to run it, if you're still interested.

https://github.com/ondrejdyck/Foliate_to_Obsidian_Highlight_Importer

**Reply by starise** [reply]
Sure, thanks for sharing Ondrej, much appreciated.
